test/test/DottyDocTests.scala

```diff
@@ -0,0 +1,506 @@
+package test
+
+import dotty.tools.dotc.Compiler
+import dotty.tools.dotc.core.Phases.Phase
+import dotty.tools.dotc.typer.FrontEnd
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.ast.Trees._
+
+/** All tests need to extend this trait and define `source` and `assertion`
+ *
+ * Instances also need to be placed in the `DottyDocTests::tests` sequence to
+ * be included in the run
+ */
+trait DottyDocTest extends DottyTest { self =>
+  /** Source code in string format */
+  def source: String
+
+  /** A test to run against the resulting code */
+  def assertion: PartialFunction[Tree[Untyped], Unit]
+
+  private def defaultAssertion: PartialFunction[Tree[Untyped], Unit] = {
+    case x => assert(false, "Couldn't match resulting AST to expected AST in: " + x.show)
+  }
+
+  private def compiler(assertion: PartialFunction[Tree[Untyped], Unit]) = new Compiler {
```

Comment:
DottyTest already has a similar method `compilerWithChecker(checkAfterPhase: String, source:String)(assertion:(tpd.Tree, Context) => Unit)`
It would be better to reuse it instead of re-implementing.